assignments/assignment7/faulty-oops.md

```diff
@@ -0,0 +1,77 @@
+x17: 0000000000000000 x16: 0000000000000000 x15: 0000000000000000
+x14: 0000000000000000 x13: 0000000000000000 x12: 0000000000000000
+x11: 0000000000000000 x10: 0000000000000000 x9 : 0000000000000000
+x8 : 0000000000000000 x7 : 0000000000000000 x6 : 0000000000000000
+x5 : 0000000000000001 x4 : ffffffc0006f7000 x3 : ffffffc008d23df0
+x2 : 0000000000000012 x1 : 0000000000000000 x0 : 0000000000000000
+Call trace:
+ faulty_write+0x14/0x20 [faulty]
+ ksys_write+0x68/0x100
+ __arm64_sys_write+0x20/0x30
+ invoke_syscall+0x54/0x130
+ el0_svc_common.constprop.0+0x44/0xf0
+ do_el0_svc+0x40/0xa0
+ el0_svc+0x20/0x60
+ el0t_64_sync_handler+0xe8/0xf0
+ el0t_64_sync+0x1a0/0x1a4
+Code: d2800001 d2800000 d503233f d50323bf (b900003f) 
+---[ end trace fb0b1365aae11947 ]---
+```
+
+_The above error was due to a **null pointer dereference** which resulted in a kernel oops, eventually leading to rebooting the kernel_
+
+## Analysing the generation of error
+- The error was generated in response to an echo command shown below.
+- ```echo “hello_world” > /dev/faulty```
+- Prior to running this, the system had already loaded the faulty.ko module and had created the faulty node in /dev.
+- The ```echo``` command invokes a write operation of the faulty kernel module.
+- After investigating the faulty.c file_ops, the faulty_write method contains an inherent bug which does a null pointer dereference as shown below on line 3.
+```
+   1. ssize_t faulty_write(struct file *filp, const char __user *buf, size_t count, loff_t *pos) {
+   2.   /* make a simple fault by dereferencing a NULL pointer */
+   3.   *(int *)0 = 0;
+   4.   return 0;
+   5.  }
+```  
+
+## Analysing the effect
+- The invocation of the error caused a _kernel oops_. 
+- At this stage the kernel terminates all the user-space processes, unmounts the file system and starts to obtain debug information about the error.
+- This is what is displayed on the terminal regarding the error.
+- If the kernel is unable to recover, the system is _rebooted_, which did happen in our case.
+
+## Analysing the debug message
+- The debug log of the oops shows the cause of panic and the kernel state as well as the userspace process state.
+- The memory fault registers and their state is displayed in the **Mem abort info** and  **Data abort info** section.
+- A brief about the userspace process memory state that caused the error is also shown in the **user pgtable** section. The page size, virtual address fields used are shown.
+- Also, what modules were loaded duuring the occurance of this panic is also provided in **Modules linked in**.
+- Finally, a detailed call trace showing what values were present at the process stack _(program counter(**pc**), link register (**lr**), stack pointer (**sp**))_ also showning **faulty_write+0x14/0x20 [faulty]** was the location of the last function call when the error happened justifies the resoning of the null pointer dereference.
```

Comment:
The null pointer reference occurred 0x14 bytes into a function `faulty_write` of length 0x20 bytes.